typescript/ws-protocol/src/types.ts

```diff
@@ -53,13 +53,19 @@ export type Unsubscribe = {
   op: "unsubscribe";
   subscriptionIds: SubscriptionId[];
 };
-
-export type ClientChannel = {
+type None<T> = { [K in keyof T]?: never };
+type BothOrNone<T1, T2> = (None<T1> & None<T2>) | (T1 & T2);
```

Comment:
I might go with `type ClientChannelSchemaInfo = { schema: string, schemaEncoding: string } | { schema?: undefined, schemaEncoding?: undefined }` instead of the generic stuff for simplicity.